src/resultset/resultset.c

```diff
@@ -206,23 +206,44 @@ void ResultSet_CachedExecution(ResultSet *set) {
 }
 
 void ResultSet_Reply(ResultSet *set) {
-	if(set->header_emitted) {
-		// If we have emitted a header, set the number of elements in the preceding array.
-		RedisModule_ReplySetArrayLength(set->ctx, set->recordCount);
-	} else if(set->header_emitted == false && set->columns != NULL) {
-		ASSERT(set->recordCount == 0);
-		// Handle the edge case in which the query was intended to return results, but none were created.
-		_ResultSet_ReplyWithPreamble(set, NULL);
-		RedisModule_ReplySetArrayLength(set->ctx, 0);
+	int nsections = 1; // queries that don't emit data will only emit statistics
+	bool require_footer = false;
+
+	if(set->columns != NULL) {
+		uint graph_version = QueryCtx_GetGraphVersion();
+		require_footer = (graph_version != GRAPH_VERSION_MISSING &&
+						  graph_version != GraphContext_GetVersion(set->gc) &&
+						  set->format == FORMATTER_COMPACT);
+
+		// the query emits data
+		if(set->header_emitted) {
+			// if we have emitted a header,
+			// set the number of elements in the preceding array
+			RedisModule_ReplySetArrayLength(set->ctx, set->recordCount);
+		} else {
+			ASSERT(set->recordCount == 0);
+			// handle the edge case in which the query was intended to
+			// return results, but none were created
+			_ResultSet_ReplyWithPreamble(set, NULL);
+			RedisModule_ReplySetArrayLength(set->ctx, 0);
+		}
+		nsections = (require_footer) ? 4 : 3;
+		RedisModule_ReplySetArrayLength(set->ctx, nsections);
 	} else {
 		// Queries that don't emit data will only emit statistics
 		RedisModule_ReplyWithArray(set->ctx, 1);
 	}
 
-	/* Check to see if we've encountered a run-time error.
-	 * If so, emit it as the last top-level response. */
+	// RedisModule_ReplyWithArray(set->ctx, nsections);
+
+	/* check to see if we've encountered a run-time error
+	 * if so, emit it as the last top-level response */
 	if(ErrorCtx_EncounteredError()) ErrorCtx_EmitException();
-	else _ResultSet_ReplayStats(set->ctx, set); // Otherwise, the last response is query statistics.
+	else _ResultSet_ReplayStats(set->ctx, set); // otherwise, the last response is query statistics
+
+	if(require_footer) {
+		set->formatter->EmitFooter(set->ctx, set->gc);
+	}
```

Comment:
The idea here was to pass a `map` data type to the formatter: `set->formatter->EmitFooter(set->ctx, map);`
where the map contains all of the information we would like to place in the footer.